src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java

```diff
@@ -717,15 +721,39 @@ public InstancesInfo getApplicationInstances(CloudApplication application) {
     public List<CloudApplication> getApplications() {
         List<CloudApplication> applications = fetchListWithAuxiliaryContent(this::getApplicationResources,
                                                                             this::zipWithAuxiliaryApplicationContent);
-        return addMetadataIfNotEmpty(applications);
+        return addMetadataIfNotEmpty(addStacksToCloudApplications(applications));
+    }
+
+    private List<CloudApplication> addStacksToCloudApplications(List<CloudApplication> cloudApps) {
+        return cloudApps.stream()
+                        .map(this::addStackToCloudApplication)
+                        .collect(Collectors.toList());
+    }
+
+    private CloudApplication addStackToCloudApplication(CloudApplication application) {
+        UUID cloudStackId = application.getStaging()
+                                       .getStackId();
+        CloudStack cloudStack = retrieveCloudStack(cloudStackId, () -> ImmutableRawCloudStack.of(getStackResource(cloudStackId).block())
+                                                                                             .derive());
+        return copyCloudApplicationWithStacks(application, cloudStack).derive();
+    }
+
+    private CloudStack retrieveCloudStack(UUID rawCloudStackId, Supplier<CloudStack> cloudStack) {
+        return cloudStackCache.getCloudStack(rawCloudStackId, cloudStack);
+    }
+
+    private ImmutableCloudApplication copyCloudApplicationWithStacks(CloudApplication cloudApp, CloudStack cloudStack) {
+        return ImmutableCloudApplication.copyOf(cloudApp)
+                                        .withStaging(ImmutableStaging.copyOf(cloudApp.getStaging())
+                                                                     .withStackName(cloudStack.getName()));
     }
 
     @Override
     public List<CloudApplication> getApplicationsByMetadataLabelSelector(String labelSelector) {
         Map<String, Metadata> applicationsMetadata = getApplicationsMetadataByLabelSelector(labelSelector);
-        List<CloudApplication> cloudApplications = fetchListWithAuxiliaryContent(() -> getApplicationResourcesByNamesInBatches(applicationsMetadata.keySet()),
-                                                                                 this::zipWithAuxiliaryApplicationContent);
-        return addMetadata(cloudApplications, applicationsMetadata);
+        return addMetadata(addStacksToCloudApplications(fetchListWithAuxiliaryContent(() -> getApplicationResourcesByNamesInBatches(applicationsMetadata.keySet()),
```

Comment:
Why is this inlined? 3 function calls is hard to read. Split them into calling and assignment.